functions/go/set-namespace/namespace_transformer.go

```diff
@@ -71,19 +90,63 @@ func (p *plugin) Transform(m resmap.ResMap) error {
 		if err != nil {
 			return err
 		}
-		if err = p.roleBindingHack(r, sal); err != nil {
+		if err = p.roleBindingHack(r); err != nil {
+			return err
+		}
+		if err = p.updateDependsOnAnnotation(r); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+// set namespace if input matches <group>/namespaces/<namespace>/<kind>/<name>
+// returns (string, bool) where the bool indicates if namespace was set
+func (p *plugin) setDependsOnNamespace(dependsOn string) (string, bool) {
+	if !namespacedResourcePattern.MatchString(dependsOn) {
+		return dependsOn, false
+	}
+	segments := strings.Split(dependsOn, "/")
+	rk := resourceKey{
+		Group: segments[groupIdx],
+		Kind:  segments[kindIdx],
+		Name:  segments[nameIdx],
+	}
+	// Only update namespace if the referenced resource is included in input
+	if !p.inputResourceLookup.HasResource(rk) {
+		return dependsOn, false
+	}
+	segments[namespaceIdx] = p.Namespace
+	dependsOn = strings.Join(segments, "/")
+	return dependsOn, true
+}
+
+// updateDependsOnAnnotation updates the namespace for the depends-on annotation
+// if the annotation is for a namespaced resource. The expected syntax for a
+// namespaced resource is <group>/namespaces/<namespace>/<kind>/<name>
+func (p *plugin) updateDependsOnAnnotation(r *resource.Resource) error {
+	annotations := r.GetAnnotations()
+	dependsOn, ok := annotations[dependsOnAnnotation]
+	if !ok {
+		return nil
+	}
+	dependsOn, ok = p.setDependsOnNamespace(dependsOn)
+	if !ok {
+		return nil
+	}
+	annotations[dependsOnAnnotation] = dependsOn
+	if err := r.SetAnnotations(annotations); err != nil {
+		return err
+	}
+	return nil
+}
+
 // roleBindingHack extends the behavior of the upstream kustomize filter
 // This extension adds the following behavior:
 // Given a ServiceAccount is present in the ResourceList, additionally set the
 // namespace for that ServiceAccount where it is referenced in the "subjects"
 // field of a RoleBinding or ClusterRoleBinding
-func (p *plugin) roleBindingHack(r *resource.Resource, sal serviceAccountLookup) error {
+func (p *plugin) roleBindingHack(r *resource.Resource) error {
```

Comment:
As Sunil suggested earlier, let's avoid using `hack` in the method name.

Reply:
Sounds good, I can update this

However, just as a reference this method is intended as an extension of the behavior of the [upstream roleBindingHack](https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/namespace/namespace.go#L101)